lib/Sema/CSSolver.cpp

```diff
@@ -1980,6 +1980,18 @@ bool ConstraintSystem::solveSimplified(
       auto locator = disjunction->getLocator();
       assert(locator && "remembered disjunction doesn't have a locator?");
       DisjunctionChoices.push_back({locator, index});
+
+      // Implicit unwraps of optionals are worse solutions than those
+      // not involving implicit unwraps.
+      if (!locator->getPath().empty()) {
+        auto kind = locator->getPath().back().getKind();
+        if (kind == ConstraintLocator::ImplicitlyUnwrappedDisjunctionChoice ||
+            kind == ConstraintLocator::DynamicLookupResult) {
+          assert(index == 0 || index == 1);
+          if (index == 1)
```

Comment:
Maybe we should add additional locator step to each of the choices here so it's easier to identify which one is unwrap?

Reply:
I'm not sure I know what you mean. These are both forced unwraps.

If you're concerned that we'll add more locators that indicate they are for a disjunction for a forced unwrap, and that despite my assert and comment we'll still fail to update this code, then we could consider doing this in another way.

I had originally thought I'd track these as they are created and perform a lookup here but thought that may end up being overkill and not necessarily any less error prone as far as future code modifications are concerned.

Reply:
Sorry, I mean for each of the choices in this disjunction we could use an extended locator which identifies constraint that bindings to `T?` and the one which bindings to `T` explicitly instead of relying on indexes so we'd have something like `<<expr>> -> implicitly unwrapped disjunction choice -> optional binding` and `<<expr>> -> implicitly unwrapped disjunction choice -> underlying type binding`

Reply:
I think that could ultimately make other things more brittle. Currently `getDisjunctionChoice()` asserts that the locator you are looking up is in the set of remembered choices. This was an important check in moving to the new IUO implementation - at least a few times it caught bugs in my code. In order to do what you're suggesting we'd have to add a way to ask whether one or the other of these locators exists, and ensure that we assert appropriately at the use site if we expect one or the other to exist and neither does. I agree that using the numeric index here isn't ideal, but I don't think the alternative is actually better.